src/libmeasurement_kit/ooni/captive_portal.cpp

```diff
@@ -0,0 +1,231 @@
+                result["failure"] = "unknown_error";
+            } else {
+                result["actual_body"] = response->body;
+                result["actual_status"] = response->status_code;
+                // all tests check status or body but never both or neither
+                result["actual_body"] = response->body;
+                bool unfiltered = false;
+                if (input.count("body")) {
+                    unfiltered = !!(input.at("body") == response->body);
+                } else if (input.count("status")) {
+                    // Note: `std::stoul` can throw but the status is set by
+                    // us, therefore we do not bother with checking exceptions
+                    unfiltered = !!(std::stoul(input.at("status")) ==
+                                    response->status_code);
+                } else {
+                    // Should really not happen, no need to be specific
+                    logger->warn("unexpected input format");
+                    result["failure"] = "unknown_error";
+                }
+                logger->info("%s unfiltered: %s", input.at("name").c_str(),
+                             unfiltered ? "yes" : "no");
+                result["result"] = unfiltered;
+            }
+            (*entry)["vendor_http_tests"].push_back(result);
+            done_cb(err);
+        };
+    };
+
+    std::vector<Continuation<Error>> continuations;
+    // Note: regarding `input` we're making copies on purpose for
+    // robustness. We can be more smart using move semantic, but it
+    // doesn't seem we need to optimize that much (it's small).
+    for (auto input : gen_http_inputs()) {
+        logger->info("setting up %s", input.at("name").c_str());
+        options["http/url"] = input.at("url");
+        std::string body;
+        http::Headers headers;
+        if (input.count("ua")) {
+            headers = {{"User-Agent", input.at("ua")}};
+        } else {
+            headers = constants::COMMON_CLIENT_HEADERS;
+        }
+
+        // Note: the following callback uses `[=]`, which means that the
+        // options are copied in each run. Thus, each test is getting its
+        // right URL to connect to and check.
+        continuations.push_back([=](Callback<Error> done_cb) {
+            templates::http_request(entry, options, headers, body,
+                                    http_cb(input, done_cb), reactor, logger);
+        });
+    }
+    mk::parallel(continuations, all_done_cb, 3);
+}
+
+// this is the only test that doesn't follow the pattern of those above.
+// this hostname should always resolve to this IP.
+void dns_msft_ncsi(Var<Entry> entry, Callback<Error> done_cb, Settings options,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    std::string hostname = "dns.msftncsi.com";
+    // Note: we're setting the nameserver to empty, which is going to work
+    // as long as we're using the `system` DNS resolver
```

Comment:
Is there a way to force system DNS resolver? What happens if in the future the MK default changes?